source/common/quic/envoy_quic_client_connection.h

```diff
@@ -68,24 +100,38 @@ class EnvoyQuicClientConnection : public quic::QuicConnection,
   }
 
   // Register file event and apply socket options.
-  void setUpConnectionSocket(OptRef<PacketsToReadDelegate> delegate);
+  void setUpConnectionSocket(Network::ConnectionSocket& connection_socket,
+                             OptRef<PacketsToReadDelegate> delegate);
 
   // Switch underlying socket with the given one. This is used in connection migration.
   void switchConnectionSocket(Network::ConnectionSocketPtr&& connection_socket);
 
+  // Potentially trigger migration.
+  void OnPathDegradingDetected() override;
+
+  void onPathValidationSuccess(std::unique_ptr<quic::QuicPathValidationContext> context);
+
+  void onPathValidationFailure(std::unique_ptr<quic::QuicPathValidationContext> context);
+
 private:
   EnvoyQuicClientConnection(const quic::QuicConnectionId& server_connection_id,
                             quic::QuicConnectionHelperInterface& helper,
                             quic::QuicAlarmFactory& alarm_factory,
                             const quic::ParsedQuicVersionVector& supported_versions,
                             Event::Dispatcher& dispatcher,
-                            Network::ConnectionSocketPtr&& connection_socket);
+                            Network::ConnectionSocketPtr&& connection_socket,
+                            const envoy::config::core::v3::QuicProtocolOptions& protocol_config);
 
   void onFileEvent(uint32_t events);
 
+  void maybeMigratePort();
+
   OptRef<PacketsToReadDelegate> delegate_;
   uint32_t packets_dropped_{0};
   Event::Dispatcher& dispatcher_;
+  const envoy::config::core::v3::QuicProtocolOptions& protocol_config_;
+  Network::ConnectionSocketPtr probing_socket_;
```

Comment:
Can this go into EnvoyQuicPathValidationContext? Its life time should be the same as writer.

Reply:
Since we can't really bind the reference of this socket in OnFileEvent, we need a way to read from this probing socket when file event is triggered.
I guess we could potentially call GetPathValidationContext() to fetch the socket? Though I think placing it in the connection might be easier.